cmd/mdatagen/internal/templates/metrics.go.tmpl

```diff
@@ -140,6 +143,72 @@ func (m *metric{{ $name.Render }}) recordDataPoint(start pcommon.Timestamp, ts p
     {{- end }}
 }
 
+// very similar to a normal data point record but if we detect that the original
+// set of attributes has been configured by the user we use this function to
+// raggregate results before emitting
+func (m *metric{{ $name.Render }}) reAggAndRecordDataPoint(start pcommon.Timestamp, ts pcommon.Timestamp, val {{ $metric.Data.MetricValueType.BasicType }}
```

Comment:
If we do it on the recording time, we don't need a separate function. We can just optionally append datapoints keeping the number of aggregated datapoints in a separate slice as part of `metric{{ $name.Render }}` struct. That number can be used at the emit time in the following way

```
func (m *metric{{ $name.Render }}) emit(metrics pmetric.MetricSlice) {
	if m.config.Enabled && m.data.{{ $metric.Data.Type }}().DataPoints().Len() > 0 {
		if m.aggrStrategy == "avg" {
			for i, aggregatedCount := m.aggregatedDatapoints {
				m.DataPoints().At(i).Gauge().SetValue(m.DataPoints().At(i).Gauge().Value() / aggregatedCount)
			}
		}
		m.updateCapacity()
		m.data.MoveTo(metrics.AppendEmpty())
		m.init()
	}
}
```